cabal-install/src/Distribution/Client/Init/Types.hs

```diff
@@ -282,13 +289,23 @@ mkLiterate _ hs = hs
 -- -------------------------------------------------------------------- --
 -- Interactive prompt monad
 
+type PromptIO = ReaderT (Data.IORef.IORef SessionState) IO
+
+runPromptIO :: PromptIO a -> IO a
+runPromptIO pio =
+  (Data.IORef.newIORef _newSessionState) >>= (runReaderT pio)
+
 newtype PurePrompt a = PurePrompt
-  { _runPrompt
-      :: NonEmpty String
-      -> Either BreakException (a, NonEmpty String)
+  { _runPromptState
+      :: (NonEmpty String, SessionState)
+      -> Either BreakException (a, (NonEmpty String, SessionState))
   }
   deriving (Functor)
 
+_runPrompt :: PurePrompt a -> NonEmpty String -> Either BreakException (a, NonEmpty String)
+_runPrompt act args =
+  (fmap (\(a, (s, _)) -> (a, s))) (_runPromptState act (args, _newSessionState))
```

Comment:
No need in the outer parenthesis. Splitting may help reading.
```suggestion
  fmap
    (\(a, (s, _)) -> (a, s)))
    (_runPromptState act (args, _newSessionState)
```

Again, plenty of underscores...